### PR TITLE
Refactor: kv&lease store

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

based on #199 

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline/src/storage/storage_api.rs [hunk 1]
```diff
@@ -4,22 +4,6 @@ use super::ExecuteError;
 
 /// The Stable Storage Api
 pub trait StorageApi: Send + Sync + 'static + std::fmt::Debug {
-    /// Insert key-value pair into the underlying storage engine
-    ///
-    /// # Errors
-    ///
-    /// if error occurs in storage, return `Err(error)`
-    fn insert<K, V>(
```
Comment (Phoenix500526) [inline xline/src/storage/storage_api.rs:12]: Why do you remove this method?

File: xline/src/storage/storage_api.rs [hunk 2]
```diff
@@ -50,15 +34,20 @@ pub trait StorageApi: Send + Sync + 'static + std::fmt::Debug {
...
-    /// Delete key-value pair from storage
+    /// Commit a batch of write operations
+    /// If sync is true, the write will be flushed from the operating system
+    /// buffer cache before the write is considered complete. If this
+    /// flag is true, writes will be slower.
     ///
     /// # Errors
     ///
     /// if error occurs in storage, return `Err(error)`
-    fn delete<K>(&self, table: &'static str, key: K, sync: bool) -> Result<(), ExecuteError>
```
Comment (Phoenix500526) [inline xline/src/storage/storage_api.rs:58]: Same as above

File: xline/src/storage/auth_store/store.rs [hunk 3]
```diff
@@ -99,10 +101,39 @@ where
         }
     }
 
-    /// generate a next revision and save it into backend
-    fn commit_revision(&self) -> Result<(), ExecuteError> {
-        let _ignore = self.revision.next();
-        self.backend.save_revision(&self.revision)
+    /// generate a next revision and create a write operation to persistent it
+    fn op_inc_and_put_revision(&self) -> WriteOperation {
```
Comment (Phoenix500526) [inline xline/src/storage/auth_store/store.rs:105]: I think we shouldn't expose some lower details, like `WriteOperation`,  to the upper layer. This goes against the law of Demeter. If we need to modify the storage engine layer for some reason in the future, this change will diffuse to the upper layer, like the Store layer and its upper layer.

File: xline/src/storage/db.rs [hunk 4]
```diff
@@ -104,22 +95,35 @@ where
...
-        Ok(())
+            .write_batch(ops, true)
+            .map_err(|e| ExecuteError::DbError(format!("Failed to reset database, error: {e}")))
     }
 
-    fn write_batch(&self, wr_ops: Vec<WriteOperation>, sync: bool) -> Result<(), ExecuteError> {
-        self.engine
-            .write_batch(wr_ops, sync)
-            .map_err(|e| ExecuteError::DbError(format!("Failed to write batch, error: {e}")))
+    fn put_op(&self, propose_id: &ProposeId, op: WriteOp) {
```
Comment (Phoenix500526) [inline xline/src/storage/db.rs:110]: `put_op` is a little bit confusing. Actually, we can draw a conclusion from context that `put_op` means to store a `WriteOp` into the buffer. But `put_op` looks like to do a put operation. How about `buffer_op`? We can tell the buffer operations and the storage operations apart. And `buffer_op` implicitly indicates that it only buffers some operations, and it should be flushed.

File: curp/src/cmd.rs [hunk 5]
```diff
@@ -101,7 +101,7 @@ impl ConflictCheck for u32 {
 /// Command executor which actually executes the command.
 /// It is usually defined by the protocol user.
 #[async_trait]
-pub trait CommandExecutor<C>: Sync + Send + Clone + std::fmt::Debug
```
Comment (markcty) [inline curp/src/cmd.rs:104]: What's the purpose of `Clone`. I thought ce is meant to be shared by `Arc` not cloning.

File: xline/src/storage/db.rs [hunk 6]
```diff
@@ -104,22 +95,35 @@ where
...
+        let mut buffer = self.buffer.lock();
+        if let Some(ops) = buffer.get_mut(propose_id) {
+            ops.push(op);
+        } else {
+            let _ignore = buffer.insert(propose_id.clone(), vec![op]);
+        }
+    }
```
Comment (markcty) [inline xline/src/storage/db.rs:117]:
```suggestion
        buffer.or_insert_with(|| vec![]).push(op);
```

Comment (themanforfree) [reply xline/src/storage/db.rs:117]: it should be `buffer.entry(propose_id.clone()).or_insert_with(|| vec![]).push(op)`, if change to this, `propose_id` will be cloned every time

File: curp/src/cmd.rs (hunk 5 above)
Comment (themanforfree) [reply curp/src/cmd.rs:104]: So I removed `Clone`

File: xline/src/storage/db.rs (hunk 4 above)
Comment (themanforfree) [reply xline/src/storage/db.rs:110]: Done

File: xline/src/storage/auth_store/store.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/storage/auth_store/store.rs]: Fixed

File: xline/src/storage/storage_api.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/storage_api.rs:12]: we didn't need it

File: xline/src/storage/storage_api.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/storage/storage_api.rs:58]: Same above

File: curp/src/cmd.rs (hunk 5 above)
Comment (markcty) [reply curp/src/cmd.rs:104]: Sry.

**Review by markcty — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**